llvm/lib/Target/AArch64/SVEInstrFormats.td

```diff
@@ -3312,14 +3312,23 @@ multiclass sve_fp_z2op_p_zd_d<bit U, string asm, string int_op, SDPatternOperato
   def : SVE_1_Op_PassthruUndefZero_Pat<nxv2i64, ir_op, nxv2i1, nxv2f64, !cast<Instruction>(NAME # _DtoD)>;
 }
 
-multiclass sve_fp_z2op_p_zd_c<bit U, string asm> {
+multiclass sve_fp_z2op_p_zd_c<bit U, string asm, string int_op, SDPatternOperator ir_op> {
   def _HtoH : sve_fp_z2op_p_zd<{ 0b011001, U }, asm, ZPR16, ZPR16>;
   def _StoH : sve_fp_z2op_p_zd<{ 0b011010, U }, asm, ZPR32, ZPR16>;
   def _StoS : sve_fp_z2op_p_zd<{ 0b101010, U }, asm, ZPR32, ZPR32>;
   def _StoD : sve_fp_z2op_p_zd<{ 0b111000, U }, asm, ZPR32, ZPR64>;
   def _DtoS : sve_fp_z2op_p_zd<{ 0b111010, U }, asm, ZPR64, ZPR32>;
   def _DtoH : sve_fp_z2op_p_zd<{ 0b011011, U }, asm, ZPR64, ZPR16>;
   def _DtoD : sve_fp_z2op_p_zd<{ 0b111011, U }, asm, ZPR64, ZPR64>;
+
+  def : SVE_3_Op_UndefZero_Pat<nxv4f32, !cast<SDPatternOperator>(int_op # _f32i64), nxv4f32, nxv2i1, nxv2i64, !cast<Instruction>(NAME # _DtoS)>;
+  def : SVE_3_Op_UndefZero_Pat<nxv2f64, !cast<SDPatternOperator>(int_op # _f64i32), nxv2f64, nxv2i1, nxv4i32, !cast<Instruction>(NAME # _StoD)>;
+  def : SVE_3_Op_UndefZero_Pat<nxv8f16, !cast<SDPatternOperator>(int_op # _f16i32), nxv8f16, nxv4i1, nxv4i32, !cast<Instruction>(NAME # _StoH)>;
+  def : SVE_3_Op_UndefZero_Pat<nxv8f16, !cast<SDPatternOperator>(int_op # _f16i64), nxv8f16, nxv2i1, nxv2i64, !cast<Instruction>(NAME # _DtoH)>;
+
+  def : SVE_1_Op_PassthruUndefZero_Pat<nxv8f16, ir_op, nxv8i1,nxv8i16, !cast<Instruction>(NAME # _HtoH)>;
```

Comment:
`AArch64{s,u}cvtf_mt` already has Undef patterns for these types - shouldn't we need some use of `UseUnaryUndefPseudos` on the existing patterns to avoid collision?

As discussed, this node is how all builtins that map to `int_aarch64_sve_{u,s}_cvtf`are lowered.

Reply:
The patterns for those `_ZPmZ` instructions contain `undef` whereas the patterns for the corresponding `_ZPzZ`  instructions contain `SVEDup0Undef`. The latter is a `ComplexPattern` with higher (even the highest!) priority, it will be checked first, if enabled by the predicate.

cf. https://github.com/llvm/llvm-project/blob/204333323db85282a6cf7ec0d49c144c73d730cb/llvm/utils/TableGen/DAGISelEmitter.cpp#L100

Reply:
I see, thanks.

Though I thought we were trying to stay away from relying on pattern complexity - if not, would this make the existing use of `UseUnaryUndefPseudos` unnecessary?

Reply:
> Though I thought we were trying to stay away from relying on pattern complexity

Yeah, when there's doubt, sure, but for a clear, well-understood cases like `undef` vs. a `SVEDup0Undef` all else being identical I see not problem.

>  - if not, would this make the existing use of `UseUnaryUndefPseudos` unnecessary?

Either that, or guard more patterns with `UseUnaryUndefPseudos`. The issue is how (and whether) to disambiguate between, for example, patterns like these:

```
 dag PatternToMatch = (nxv8f16 (AArch64scvtf_mt (nxv8i1 (SVEAllActive:Op1)), nxv8i16:$Op2, nxv8f16:$Op3));
```

and

```
  dag PatternToMatch = (nxv8f16 (AArch64scvtf_mt nxv8i1:$Op1, nxv8i16:$Op2, (nxv8f16 (SVEDup0Undef))));
```

Reply:
> ` dag PatternToMatch = (nxv8f16 (AArch64scvtf_mt (nxv8i1 (SVEAllActive:Op1)), nxv8i16:$Op2, nxv8f16:$Op3));`

This stream of work seems to have assumed that something like `instCombineSVEAllActiveReplace` has taken care of this already, or am I misunderstanding something?

E.g, if we run llc directly on:
```
define <vscale x 4 x float> @test(<vscale x 4 x float> %y, <vscale x 4 x i32> %x) #0 {
entry:
  %0 = tail call <vscale x 4 x i1> @llvm.aarch64.sve.ptrue.nxv4i1(i32 31)
  %1 = tail call <vscale x 4 x float> @llvm.aarch64.sve.scvtf.nxv4f32.nxv4i32(<vscale x 4 x float> %y, <vscale x 4 x i1> %0, <vscale x 4 x i32> %x)
  ret <vscale x 4 x float> %1
}
attributes #0 = {"target-features"="+sve2p2"}
```
We'd still get 
```
	ptrue	p0.s
	movprfx	z0, z1
	scvtf	z0.s, p0/m, z1.s
	ret
```

Reply:
Or, are you referring to a case where we are passing an `undef` passthrough operand ***and*** using an all-active predicate? Which would seem to be the case after  `instCombineSVEAllActiveReplace` has acted - and seems to match `SVEAllActive` instead of `SVEDupZeroUndef`.

Reply:
Yes, here we are concerned only with `undef` in the inactive lanes. If you look at the `fabs` case there there's no `movprfx` even with SVE2.1 and I don't think we prefer ` fabs z0.s, p0/m, z0.s` over ` fabs z0.s, p0/z, z0.s` or vice versa.
https://gcc.godbolt.org/z/Mje48odeh

I though that's the case for `scvtf` (and others) , but apparently that's not the case.
In conclusion, this is a strong argument in favour of introducing more `UseUnaryUndefPseudos`.

Reply:
Yeah - though https://gcc.godbolt.org/z/rovvr9bzP  feels like a bit of an unwanted side-effect of `UseUnaryUndefPseudos`. We might need to take a look into which nodes actually need patterns for `SVEAllActive`, and which have this case folded by instcombine.

What do you think?

Reply:
I think we should have a pattern like
```
  def : Pat<(vtd (op (pg (SVEAllActive:$Op1)), vts:$Op2, vtd:$Op3)),
            (inst $Op3, $Op1, $Op2)>;
```
enabled in SVE2.2 to generate a `_ZPzZ` instruction and the same pattern that generates a `_ZPmZ_X_UNDEF` ought to be disabled.

Reply:
If there are still unresolved issues, we'd better restart this conversation.